drivers/xen/regions.c

```diff
@@ -0,0 +1,188 @@
+ *
+ * Check reg addr and reg size to be aligned to
+ * CONFIG_MMU_PAGE_SIZE. This is the requirement for arch_mem_map call.
+ *
+ * Extended regions should not overlap kernel Virtual Memory region.
+ */
+#define EXT_BLOCK(node_id, idx)						\
+	BUILD_ASSERT(DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx),		\
+		     "Extended region starting from 0 is not supported, please skip page 0"); \
+	BUILD_ASSERT(DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx) ==		\
+		     ROUND_DOWN(DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx), CONFIG_MMU_PAGE_SIZE), \
+		     "Extended region size is not aligned");		\
+	BUILD_ASSERT(DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx) ==		\
+		     ROUND_DOWN(DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx), CONFIG_MMU_PAGE_SIZE), \
+		     "Extended region address is not aligned");		\
+	BUILD_ASSERT((DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx) < CONFIG_KERNEL_VM_BASE || \
+		      DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx) >=		\
+		      CONFIG_KERNEL_VM_BASE + CONFIG_KERNEL_VM_SIZE),	\
+		     "Extended region should not overlap virtual ram");	\
+	BUILD_ASSERT(((DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx) +		\
+		      DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx)) <= CONFIG_KERNEL_VM_BASE || \
+		     (DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx) +		\
+		      DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx)) >		\
+		     CONFIG_KERNEL_VM_BASE + CONFIG_KERNEL_VM_SIZE),	\
+		     "Extended region should not overlap virtual ram");	\
+	SYS_MEM_BLOCKS_DEFINE_STATIC_WITH_EXT_BUF(			\
+		region_##idx, XEN_PAGE_SIZE,				\
+		XEN_PFN_DOWN(DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx)),	\
+		(uint8_t *)DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx));
+
+DT_FOREACH_REG(XEN_HYP_NODE, EXT_BLOCK);
+
+#define EXT_REG_POINTER(node_id, idx)					\
+	{								\
+		.block = &region_##idx,					\
+		.phys_addr = DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx),	\
+		.size = DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx),		\
+	},
+
+/*
+ * TODO: For now we allocate buffer structures for DT_NUM_REGS, when we
+ * need a range 1...DT_FOREACH_REG. Although region 0 is ignored, it still
+ * takes space for sys_block structures allocation. This requires optimization
+ */
+static struct extended_regions ext_blocks[DT_NUM_REGS(XEN_HYP_NODE)] = {
+	DT_FOREACH_REG(XEN_HYP_NODE, EXT_REG_POINTER)
+};
+
+void *xen_region_get_pages(size_t nr_pages)
+{
+	int ret, i;
+	void *ptr;
+
+	/*
+	 * Locking ranges iteration, although each call sys_mem_blocks_alloc_contiguous
+	 * has it's own syncronisation, to avoid the situation of 2 parralel calls of
+	 * get_page and put_page on full buffers when put_page clears space in block 1
+	 * when get_page has already checked block 1. This will lead us to -ENOMEM while
+	 * block 1 space was freed already
+	 */
+	k_mutex_lock(&ext_regions_lock, K_FOREVER);
+
+	for (i = EXTENDED_REGIONS_IDX; i < DT_NUM_REGS(XEN_HYP_NODE); i++) {
+		ret = sys_mem_blocks_alloc_contiguous(ext_blocks[i].block, nr_pages, &ptr);
+		if (ret == 0) {
+			break;
+		} else if (ret && ret != -ENOMEM) {
+			LOG_ERR("Unable to allocate pages, ret = %d\n", ret);
```

Comment:
Here and in all other cases: you don't need \n for LOG_ macros.

Reply:
+